### PR TITLE
2.x Build and Pass with PHP7.2

### PR DESCRIPTION
**Pull request by tenkoma**
Refs #11346

In this Pull Request, Installed the `mcrypt` extension from pecl in the CI so that we can test it.

- added `pear config-set preferred_state snapshot`. Because pecl's mcrypt package is `snapshot`, is not `stable` https://pecl.php.net/package/mcrypt
- I changed `pecl install` to execute first. Because fails `composer require "phpunit/phpunit=$PHPUNIT"`
- remove cannot run test matrix: PHPUnit 3.7 + PHP7.2

### CODE REVIEW

**Comment by dereuromark** [comment]
Now you are getting somewhere, errors become visible:

- count(): Parameter must be an array or an object that implements Countable
- ini_set(): A session is active. You cannot change the session module's ini settings at this time

Seem to be the issues left for PHP7.2 compatibility.

**Comment by markstory** [comment]
> ini_set(): A session is active. You cannot change the session module's ini settings at this time

These are a big headache to fix.

**Comment by tenkoma** [comment]
> ini_set(): A session is active. You cannot change the session module's ini settings at this time

I didn't fail when I only ran `CakeSessionTest`.
There was a problem with other tests.
But a new error was found.

> ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

Probably related to the next ticket.

https://bugs.php.net/bug.php?id=73100

File: lib/Cake/Model/Datasource/CakeSession.php [hunk 1]
```diff
@@ -552,6 +552,12 @@ protected static function _configureSession() {
 
 		if (!empty($sessionConfig['handler'])) {
 			$sessionConfig['ini']['session.save_handler'] = 'user';
+
+			// In PHP7.2.0+ session.save_handler can't be set to 'user' by the user.
+			// https://github.com/php/php-src/commit/a93a51c3bf4ea1638ce0adc4a899cb93531b9f0d
+			if (version_compare(PHP_VERSION, '7.2.0', '>=')) {
+				unset($sessionConfig['ini']['session.save_handler']);
+			}
```
Comment (tenkoma) [inline lib/Cake/Model/Datasource/CakeSession.php]: I referred to the following code. However, I think it is PHP 7.2 or higher that this setting cannot be done

https://github.com/cakephp/cakephp/blob/8b3905ad6f1d2f17d20d495e12e9947bd69644fb/src/Network/Session.php#L111-L114

- session.c on PHP 7.1 branch: https://github.com/php/php-src/blob/PHP-7.1/ext/session/session.c#L552-L577
- session.c on PHP 7.2 branch: https://github.com/php/php-src/blob/PHP-7.2/ext/session/session.c#L559-L563

Comment (markstory) [reply lib/Cake/Model/Datasource/CakeSession.php]: I think you have the correct code here. There is an open issue for database sessions in PHP7.2 (#11628)

File: lib/Cake/Console/Command/Task/ControllerTask.php [hunk 2]
```diff
@@ -328,7 +328,7 @@ public function bakeActions($controllerName, $admin = null, $wannaUseSession = t
  * @param array $components Components to use in controller
  * @return string Baked controller
  */
-	public function bake($controllerName, $actions = '', $helpers = null, $components = null) {
+	public function bake($controllerName, $actions = '', $helpers = array(), $components = array()) {
```
Comment (chinpei215) [inline lib/Cake/Console/Command/Task/ControllerTask.php]: Why did you change this signature?

Comment (tenkoma) [reply lib/Cake/Console/Command/Task/ControllerTask.php]: It is a change to correct errors that occur in the following lines.

https://github.com/cakephp/cakephp/blob/2.x/lib/Cake/Console/Templates/default/classes/controller.ctp#L51

But it is not BC with existing applications.
Make `null` be converted to `array()`.

File: lib/Cake/Console/Command/Task/ControllerTask.php [hunk 3]
```diff
@@ -328,9 +328,15 @@ public function bakeActions($controllerName, $admin = null, $wannaUseSession = t
  * @param array $components Components to use in controller
  * @return string Baked controller
  */
-	public function bake($controllerName, $actions = '', $helpers = array(), $components = array()) {
+	public function bake($controllerName, $actions = '', $helpers = null, $components = null) {
 		$this->out("\n" . __d('cake_console', 'Baking controller class for %s...', $controllerName), 1, Shell::QUIET);
 
+		if (is_null($helpers)) {
```
Comment (markstory) [inline lib/Cake/Console/Command/Task/ControllerTask.php]: We usually favor `$helpers === null` as it uses fewer function calls.
